src/Fetch.js

```diff
@@ -295,303 +57,217 @@ function fetchXHR(fetch, onsuccess, onerror, onprogress, onreadystatechange) {
-        else successcb?.(fetch);
-      });
-    };
-    fetchCacheData(Fetch.dbInstance, fetch, xhr.response, storeSuccess, storeError);
-  };
+      if (onerror) {{{ makeDynCall('vp', 'onerror') }}}(fetchHandle);
+    });
 
-  var performCachedXhr = (fetch, xhr, e) => {
-#if FETCH_DEBUG
-    dbg(`fetch: starting (cached) XHR: ${e}`);
-#endif
-    fetchXHR(fetch, cacheResultAndReportSuccess, reportError, reportProgress, reportReadyStateChange);
-  };
+    HEAPU16[fetchHandle + {{{ C_STRUCTS.emscripten_fetch_t.readyState }}} >> 1] = 1; // XMLHttpRequest OPENED, kept for compatibility
 
-  var requestMethod = UTF8ToString(fetch_attr + {{{ C_STRUCTS.emscripten_fetch_attr_t.requestMethod }}});
-  var fetchAttrReplace = !!(fetchAttributes & {{{ cDefs.EMSCRIPTEN_FETCH_REPLACE }}});
-  var fetchAttrPersistFile = !!(fetchAttributes & {{{ cDefs.EMSCRIPTEN_FETCH_PERSIST_FILE }}});
-  var fetchAttrNoDownload = !!(fetchAttributes & {{{ cDefs.EMSCRIPTEN_FETCH_NO_DOWNLOAD }}});
-  if (requestMethod === 'EM_IDB_STORE') {
-    // TODO(?): Here we perform a clone of the data, because storing shared typed arrays to IndexedDB does not seem to be allowed.
-    var ptr = {{{ makeGetValue('fetch_attr', C_STRUCTS.emscripten_fetch_attr_t.requestData, '*') }}};
-    var size = {{{ makeGetValue('fetch_attr', C_STRUCTS.emscripten_fetch_attr_t.requestDataSize, '*') }}};
-    fetchCacheData(Fetch.dbInstance, fetch, HEAPU8.slice(ptr, ptr + size), reportSuccess, reportError);
-  } else if (requestMethod === 'EM_IDB_DELETE') {
-    fetchDeleteCachedData(Fetch.dbInstance, fetch, reportSuccess, reportError);
-  } else if (!fetchAttrReplace) {
-    fetchLoadCachedData(Fetch.dbInstance, fetch, reportSuccess, fetchAttrNoDownload ? reportError : (fetchAttrPersistFile ? performCachedXhr : performUncachedXhr));
-  } else if (!fetchAttrNoDownload) {
-    fetchXHR(fetch, fetchAttrPersistFile ? cacheResultAndReportSuccess : reportSuccess, reportError, reportProgress, reportReadyStateChange);
-  } else {
-#if FETCH_DEBUG
-    dbg('fetch: Invalid combination of flags passed.');
-#endif
-    return 0; // todo: free
-  }
-  return fetch;
-#else // !FETCH_SUPPORT_INDEXEDDB
-  fetchXHR(fetch, reportSuccess, reportError, reportProgress, reportReadyStateChange);
-  return fetch;
-#endif // ~FETCH_SUPPORT_INDEXEDDB
+  // TODO:
+  // Check whether supporting fetchAttrSynchronous still makes sense
 }
 
 function fetchGetResponseHeadersLength(id) {
-  return lengthBytesUTF8(Fetch.xhrs.get(id).getAllResponseHeaders()) + 1;
+  if (Fetch.requestHandles.has(id)) {
+    var fetchObject = Fetch.requestHandles.get(id);
+    return lengthBytesUTF8(fetchObject.headers) + 1;
+  }
 }
 
 function fetchGetResponseHeaders(id, dst, dstSizeBytes) {
-  var responseHeaders = Fetch.xhrs.get(id).getAllResponseHeaders();
-  var lengthBytes = lengthBytesUTF8(responseHeaders) + 1;
-  stringToUTF8(responseHeaders, dst, dstSizeBytes);
-  return Math.min(lengthBytes, dstSizeBytes);
+  if (Fetch.requestHandles.has(id)) {
```

Comment:
How come these two function need this protection now but didn't before?  Was this just an oversight?  (i.e. should this `if` always have been here?)

Reply:
Seems like an oversight, technically we could remove the check and add comments in the C API to make sure the request is still active though.